File: src/lib/safe-redirect.ts
// Validates a `next` redirect target as a safe, same-origin relative path.
// Rejects absolute URLs and protocol-relative ("//host", "/\host") values to
// prevent open redirects. Returns the fallback when next is missing or unsafe.
export function safeNextPath(
  next: string | null | undefined,
  fallback = '/link'
): string {
  if (!next) return fallback
  if (!next.startsWith('/')) return fallback
  // Reject protocol-relative ("//") and backslash tricks ("/\") that browsers
  // may treat as a scheme-relative URL to another host.
  if (next[1] === '/' || next[1] === '\\') return fallback
  return next
}

File: djclass_overlay/common/safe_redirect.py
"""Validate a `next` redirect target as a safe, same-origin relative path.

Port of src/lib/safe-redirect.ts. Rejects absolute and protocol-relative URLs
("//host", "/\\host") to prevent open redirects.
"""


def safe_next_path(next_path: str, fallback: str = "/dashboard/") -> str:
    if not next_path:
        return fallback
    if not next_path.startswith("/"):
        return fallback
    # Reject protocol-relative ("//") and backslash tricks ("/\") that browsers
    # may treat as a scheme-relative URL to another host.
    if next_path[1:2] in ("/", "\\"):
        return fallback
    return next_path
